circuit/src/merkle_tree/native.rs

```diff
@@ -0,0 +1,183 @@
+use std::marker::PhantomData;
+
+use halo2wrong::halo2::arithmetic::FieldExt;
+use num_traits::pow;
+use std::collections::HashMap;
+
+use crate::{params::RoundParams, poseidon::native::Poseidon};
+const WIDTH: usize = 5;
+
+#[derive(Clone, Debug)]
+/// MerkleTree structure
+struct MerkleTree<F: FieldExt, P>
+where
+	P: RoundParams<F, WIDTH>,
+{
+	/// PhantomData for the params
+	_params: PhantomData<P>,
+	/// Variables with level and value to represent nodes
+	nodes: HashMap<usize, Vec<F>>,
+	/// Height of the tree
+	height: usize,
+}
+
+impl<F: FieldExt, P> MerkleTree<F, P>
+where
+	P: RoundParams<F, WIDTH>,
+{
+	/// Build a MerkleTree from given leaf nodes
+	fn build_tree(mut leaves: Vec<F>) -> Self {
+		// 0th level is the leaf level and the max level is the root level
+		let mut nodes = HashMap::new();
+		// Calculating height from number of leaves
+		let mut height = leaves.len();
+		for i in 0..height {
+			if height == 1 {
+				height = i;
+				if pow(2, i) < leaves.len() {
+					height += 1;
+				}
+				break;
+			}
+			height /= 2;
+		}
```

Comment:
Next power of 2 for leaves length can be calculated by simply calling `next_power_of_two` function on `usize`.
`height` is then `log2` of that calculated value

Reply:
But calling `next_power_of_two` wont be necessary I think, given this: https://github.com/eigen-trust/eigen-trust/pull/123#discussion_r1054276066

Reply:
Okay, thank you for the information